File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.Stack;

class Solution {
    boolean solution(String s) {
        Stack<Character> stack = new Stack<>();  
        boolean answer = false; 
        
        if(stack.isEmpty() && s.charAt(0) == ')') {
            return false;
        } else {
            for(int i = 0; i < s.length(); i++) {
                char ch = s.charAt(i);
            
                stack.push(ch);
            
                if (stack.peek() == ')') {
                    stack.pop();
                    if (stack.isEmpty() == true) {
                        return false; 
                    }
                    stack.pop();
                }
            }
        
            if (stack.size() == 0) {
                answer = true; 
            }
        }       
        
        return answer;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <stack>

using namespace std;

bool solution(string s)
{
    stack<int> st;
    
    if (s[0] == ')') {
        return false; 
    }
    
    st.push(s[0]);
    
    for (int i = 1; i < s.size(); i++) {
        if (s[i] == ')') {
            if (st.empty()) {
                return false;
            }
            st.pop();
        } else {
            st.push(s[i]);
        }
    }
    return (st.empty())? true: false;
}